src/ledger/LedgerState.cpp

```diff
@@ -0,0 +1,801 @@
+        StateHeader lhr(new LedgerHeaderReference(header, previous));
+        lhr->invalidate();
+        mParent->mHeader = lhr;
+    }
+}
+
+void
+LedgerState::mergeStateIntoRoot()
+{
+    for (auto const& state : mState)
+    {
+        state.second->invalidate();
+        if (state.second->ignoreInvalid().entry())
+        {
+            storeInDatabase(state.second);
+        }
+        else
+        {
+            deleteFromDatabase(state.second);
+        }
+    }
+}
+
+void
+LedgerState::mergeHeaderIntoRoot()
+{
+    if (mHeader)
+    {
+        storeHeaderInDatabase();
+        mRoot->setPreviousHeader(mHeader->ignoreInvalid().header());
+    }
+}
+
+void
+LedgerState::rollback()
+{
+    assert(mParent || mRoot);
+
+    if (mChild)
+    {
+        mChild->rollback();
+    }
+
+    if (mParent)
+    {
+        mParent->mChild = nullptr;
+        mParent = nullptr;
+    }
+    else
+    {
+        mRoot->hasChild(false);
+        mRoot = nullptr;
+    }
+
+    mHeader.reset();
+    mState.clear();
+    mLoadBestOfferContext.clear();
+}
+
+LedgerState::StateEntry
+LedgerState::create(LedgerEntry const& entry)
+{
+    assert(mParent || mRoot);
+    assert(!mChild);
+
+    auto key = LedgerEntryKey(entry);
+    auto ler = createHelper(entry, key);
+    mState[key] = ler;
+
+    if (entry.data.type() == OFFER)
+    {
+        OfferEntry const& offer = entry.data.offer();
+        invalidateLoadBestOfferContext(offer.selling, offer.buying);
+    }
+
+    return ler;
+}
+
+LedgerState::StateEntry
+LedgerState::createHelper(LedgerEntry const& entry, LedgerKey const& key)
+{
+    StateEntry ler;
+
+    auto iter = mState.find(key);
+    if (iter != mState.end())
+    {
+        if (iter->second->ignoreInvalid().entry())
+        {
+            throw std::runtime_error("Key already exists in memory");
+        }
+        auto newEntry = std::make_shared<LedgerEntry>(entry);
+        ler = StateEntry(new LedgerEntryReference(newEntry, nullptr));
+    }
+    else if (mParent)
+    {
+        ler = mParent->createHelper(entry, key);
+    }
+    else
+    {
+        if (loadFromDatabase(key))
+        {
+            throw std::runtime_error("Key already exists in database");
+        }
+        auto newEntry = std::make_shared<LedgerEntry>(entry);
+        ler = StateEntry(new LedgerEntryReference(newEntry, nullptr));
+    }
+    return ler;
+}
+
+LedgerState::StateEntry
+LedgerState::load(LedgerKey const& key)
+{
+    assert(mParent || mRoot);
+    assert(!mChild);
+
+    auto ler = loadHelper(key);
+    mState[key] = ler;
+
+    if (ler->entry()->data.type() == OFFER)
+    {
+        OfferEntry const& offer = ler->entry()->data.offer();
+        invalidateLoadBestOfferContext(offer.selling, offer.buying);
+    }
+
+    return ler;
+}
+
+LedgerState::StateEntry
+LedgerState::loadHelper(LedgerKey const& key)
+{
+    StateEntry ler;
+
+    auto iter = mState.find(key);
+    if (iter != mState.end())
+    {
```

Comment:
when can this happen? it is unclear to me

Reply:
The comment on line 226 gives an example of how this can happen.